### PR TITLE
Change tests to use equalToForTests rather than comparing getObjectValue

### LINKED ISSUE #3693: Fix complex map rendering in CLI

### PR DESCRIPTION
**Pull request by raghavsethi**
This is a pre-requisite to changing the JSON serialization for maps, which will fix #3693.

### CODE REVIEW

File: presto-hive/src/test/java/com/facebook/presto/hive/util/TestSerDeUtils.java [hunk 1]
```diff
@@ -288,14 +285,6 @@ private static void assertBlockEquals(Block actual, Block expected)
         assertEquals(blockToSlice(actual), blockToSlice(expected));
     }
```
Comment (raghavsethi) [inline presto-hive/src/test/java/com/facebook/presto/hive/util/TestSerDeUtils.java]: Moved this to presto-spi as this is also required in BlockSerdeUtil

File: presto-main/src/main/java/com/facebook/presto/block/BlockSerdeUtil.java [hunk 2]
```diff
@@ -39,11 +40,66 @@ public static Block readBlock(BlockEncodingSerde blockEncodingSerde, Slice slice
...
     }
+
+    public static Slice blockToSlice(Block block)
+    {
+        SliceOutput sliceOutput = new DynamicSliceOutput(1000);
+        writeBlock(sliceOutput, block.copyRegion(0, block.getPositionCount()));
+        return sliceOutput.slice();
+    }
+
+    public static class TypeAndBlock
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/block/BlockSerdeUtil.java]: This is the container for type and block information used by tests now.

File: presto-main/src/main/java/com/facebook/presto/type/TypeJsonUtils.java [hunk 3]
```diff
@@ -203,15 +207,6 @@ public static boolean canCastFromJson(Type type)
         return false;
     }
 
-    private static boolean isValidJsonObjectKeyType(Type type)
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/type/TypeJsonUtils.java]: Moved to presto-spi, as this is also required in multiple places

File: presto-main/src/test/java/com/facebook/presto/block/BlockAssertions.java [hunk 4]
```diff
@@ -33,19 +40,176 @@
...
+import static com.facebook.presto.type.TypeUtils.checkElementNotNull;
+import static java.lang.Double.isNaN;
 import static java.util.Objects.requireNonNull;
 import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.fail;
 
 public final class BlockAssertions
 {
-    private BlockAssertions()
+    private BlockAssertions() {}
```
Comment (raghavsethi) [inline presto-main/src/test/java/com/facebook/presto/block/BlockAssertions.java]: Contains a bunch of code copied from all the structural type.equalTos
Is more permissive in general that equalTo, as tests rely on null values and NaNs comparing intuitively.

File: presto-main/src/test/java/com/facebook/presto/block/BlockAssertions.java [hunk 5]
```diff
@@ -320,4 +484,44 @@ public static RunLengthEncodedBlock createRLEBlock(long value, int positionCount
         BIGINT.writeLong(blockBuilder, value);
         return new RunLengthEncodedBlock(blockBuilder.build(), positionCount);
     }
+
+    private static final class KeyWrapper
```
Comment (raghavsethi) [inline presto-main/src/test/java/com/facebook/presto/block/BlockAssertions.java]: Copied from MapType.equalTo

File: presto-main/src/test/java/com/facebook/presto/operator/aggregation/TestCovarianceSampAggregation.java [hunk 6]
```diff
@@ -46,27 +45,12 @@ protected String getFunctionName()
...
-            }
-
-            @Override
-            public int hashCode()
-            {
-                // required by checkstyle
-                return super.hashCode();
-            }
-        };
+        return covariance.covariance(constructDoublePrimitiveArray(start + 5, length), constructDoublePrimitiveArray(start, length), true);
```
Comment (raghavsethi) [inline presto-main/src/test/java/com/facebook/presto/operator/aggregation/TestCovarianceSampAggregation.java]: We don't need anonymous classes any more as our equalToForTests uses error bounds in double comparisons

File: presto-spi/src/main/java/com/facebook/presto/spi/type/JsonSerde.java [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+    {
+        String baseType = typeSignature.getBase();
+        return baseType.equals(StandardTypes.BOOLEAN) ||
+                baseType.equals(StandardTypes.BIGINT) ||
+                baseType.equals(StandardTypes.DOUBLE) ||
+                baseType.equals(StandardTypes.TIMESTAMP) ||
+                baseType.equals(StandardTypes.VARCHAR);
+    }
+
+    public static class MapEntry<K, V>
```
Comment (raghavsethi) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/JsonSerde.java]: Using this class will allow us to upgrade to the newer Jackson versions that handle maps differently.

File: presto-spi/src/main/java/com/facebook/presto/spi/type/Type.java [hunk 8]
```diff
@@ -78,7 +78,7 @@
      * {@code position}. This is the value returned to the user via the
      * REST endpoint and therefore must be JSON serializable.
      */
-    Object getObjectValue(ConnectorSession session, Block block, int position);
+    Object getJsonSerializableObjectValue(ConnectorSession session, Block block, int position);
```
Comment (raghavsethi) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/Type.java]: I think part of the reason this was being used incorrectly in tests was that the name was not descriptive enough.

Comment (electrum) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/Type.java]: This should be done in a separate commit

**Comment by electrum** [comment]
This is actually a sever side API change, not a CLI change.

File: presto-main/src/main/java/com/facebook/presto/testing/TestEquality.java [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+import static com.facebook.presto.type.TypeUtils.checkElementNotNull;
+import static java.lang.Double.isInfinite;
+import static java.lang.Double.isNaN;
+
+public final class TestEquality
+{
+    private TestEquality() {}
+
+    public static boolean equalToForTests(Type type, Block leftBlock, int leftPosition, Block rightBlock, int rightPosition)
+    {
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/testing/TestEquality.java]: This is code copied from a bunch of type.equalTo methods but is a little more permissive because tests depend on stuff like NaNs comparing intuitively.

File: presto-main/src/main/java/com/facebook/presto/testing/TestEquality.java [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+import java.util.HashMap;
+import java.util.Map;
+
+import static com.facebook.presto.testing.TestingConnectorSession.SESSION;
+import static com.facebook.presto.type.TypeUtils.checkElementNotNull;
+import static java.lang.Double.isInfinite;
+import static java.lang.Double.isNaN;
+
+public final class TestEquality
+{
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/testing/TestEquality.java]: This file is in main/testing because it is also used by presto-orc

File: presto-spi/src/main/java/com/facebook/presto/spi/type/JsonSerde.java [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+    public static boolean isValidJsonObjectKeyType(TypeSignature typeSignature)
+    {
+        String baseType = typeSignature.getBase();
+        return baseType.equals(StandardTypes.BOOLEAN) ||
+                baseType.equals(StandardTypes.BIGINT) ||
+                baseType.equals(StandardTypes.DOUBLE) ||
+                baseType.equals(StandardTypes.TIMESTAMP) ||
+                baseType.equals(StandardTypes.VARCHAR);
+    }
+}
```
Comment (raghavsethi) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/JsonSerde.java]: The MapEntry class will appear here when we move to serializing maps as lists.

**Comment by electrum** [comment]
The commit message title is too long. Please limit to ~50 characters: http://chris.beams.io/posts/git-commit/

File: presto-main/src/main/java/com/facebook/presto/testing/TestEquality.java [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+                Integer leftValuePosition = wrappedLeftMap.get(key);
+                if (leftValuePosition == null) {
+                    return false;
+                }
+                int rightValuePosition = position + 1;
+
+                if (leftMapBlock.isNull(leftValuePosition) && rightMapBlock.isNull(rightValuePosition)) {
+                    return true;
+                }
+                else if (leftMapBlock.isNull(leftValuePosition) ^ rightMapBlock.isNull(rightValuePosition)) {
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/testing/TestEquality.java]: Remove `else`

File: presto-main/src/main/java/com/facebook/presto/testing/TestEquality.java [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+                    return false;
+                }
+                else if (!equalToForTests(mapType.getValueType(), leftMapBlock, leftValuePosition, rightMapBlock, rightValuePosition)) {
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/testing/TestEquality.java]: Remove `else`

Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/testing/TestEquality.java]: write `return equalToForTests(mapType.getValueType(), leftMapBlock, leftValuePosition, rightMapBlock, rightValuePosition` instead

Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/testing/TestEquality.java]: We would like to continue iterating through the loop.

**Comment by raghavsethi** [comment]
Closing for now as this requires making breaking changes to client API.
